server/s3_bucket_metadata_cache.cc

```diff
@@ -0,0 +1,417 @@
+// class S3BucketMetadataCache
+// ************************************************************************* //
+
+S3BucketMetadataCache::S3BucketMetadataCache(
+    unsigned max_cache_size, unsigned expire_interval_sec,
+    unsigned refresh_interval_sec, std::shared_ptr<MotrAPI> s3_motr_api,
+    std::shared_ptr<S3MotrKVSReaderFactory> motr_kvs_reader_factory,
+    std::shared_ptr<S3MotrKVSWriterFactory> motr_kvs_writer_factory,
+    std::shared_ptr<S3GlobalBucketIndexMetadataFactory>
+        global_bucket_index_metadata_factory)
+    : max_cache_size(max_cache_size),
+      expire_interval_sec(expire_interval_sec),
+      refresh_interval_sec(refresh_interval_sec) {
+
+  assert(!p_bucket_metadata_cache);  // Only one instance is thought
+
+  this->s3_motr_api = s3_motr_api ? std::move(s3_motr_api)
+                                  : std::make_shared<ConcreteMotrAPI>();
+  this->motr_kvs_reader_factory =
+      motr_kvs_reader_factory ? std::move(motr_kvs_reader_factory)
+                              : std::make_shared<S3MotrKVSReaderFactory>();
+  this->motr_kvs_writer_factory =
+      motr_kvs_writer_factory ? std::move(motr_kvs_writer_factory)
+                              : std::make_shared<S3MotrKVSWriterFactory>();
+  this->global_bucket_index_metadata_factory =
+      global_bucket_index_metadata_factory
+          ? std::move(global_bucket_index_metadata_factory)
+          : std::make_shared<S3GlobalBucketIndexMetadataFactory>();
+}
+
+S3BucketMetadataCache::~S3BucketMetadataCache() = default;
+
+void S3BucketMetadataCache::remove_item(std::string bucket_name) {
+
+  assert(sorted_by_access.size() == sorted_by_update.size());
+  assert(sorted_by_access.size() == items.size());
+
+  auto map_it = items.find(bucket_name);
+  assert(items.end() != map_it);
+
+  auto* p_item = (*map_it).second.get();
+
+  sorted_by_access.erase(p_item->ptr_access);
+  sorted_by_update.erase(p_item->ptr_update);
+  items.erase(map_it);
+
+  s3_log(S3_LOG_DEBUG, "",
+         "Metadata for \"%s\" has been removed from the cache",
+         bucket_name.c_str());
+}
+
+void S3BucketMetadataCache::shrink() {
+
+  for (auto rit = sorted_by_access.rbegin(); rit != sorted_by_access.rend();
+       ++rit) {
+    auto* p_item = *rit;
+
+    if (p_item->can_remove()) {
+
+      s3_log(S3_LOG_DEBUG, "",
+             "Bucket \"%s\" is selected to be removed from metadata cache",
+             p_item->get_bucket_name().c_str());
+
+      remove_item(p_item->get_bucket_name());
+      return;
+    }
+  }
+  s3_log(S3_LOG_WARN, "", "Cannt remove any item from bucket metadata cache");
+}
+
+S3BucketMetadataCache::Item* S3BucketMetadataCache::get_item(
+    const S3BucketMetadata& src) {
+
+  const auto& bucket_name = src.get_bucket_name();
+
+  if (items.end() == items.find(bucket_name) &&
+      items.size() >= max_cache_size) {
+
+    s3_log(S3_LOG_DEBUG, src.get_request_id(), "Bucket metadata cache is full");
+    shrink();
```

Comment:
shrink seems to list of all buckets, so every fetch operation from cache is linear time and not const, see if we can make this const time

Reply:
shrink stops lookup on first entry which is ready for remove from the cache.
In most times the lookup stops on the first entry